core/src/main/java/hudson/model/Job.java

```diff
@@ -1143,6 +1148,22 @@ private HealthReport getBuildStabilityHealthReport() {
         int failCount = 0;
         int totalCount = 0;
         RunT i = getLastBuild();
+        RunT u = getLastFailedBuild();
+        if (i != null && u == null) {
+                // no failures, like ever
+                return new HealthReport(100, Messages._Job_BuildStability(Messages._Job_NoRecentBuildFailed()));
+        }
+        if (i != null && u.getNumber() <= i.getNumber()) {
+            int recentSuccessCount = _getRuns().subMap(u.getNumber() + 1, i.getNumber() + 1).size();
+            if (recentSuccessCount >= 5) {
+                return new HealthReport(100, Messages._Job_BuildStability(Messages._Job_NoRecentBuildFailed()));
+            } else {
+                // no need to load the intermediate builds, they're all successful, let's start iterating from
```

Comment:
Perhaps [this code](https://github.com/jenkinsci/jenkins/blob/1c4c6432b795f807758152d06f3f8d9a7b9bf7bb/core/src/main/java/jenkins/model/lazy/BuildReferenceMapAdapter.java#L91) works, by avoiding `unwrap`. But needs to be mechanically verified.